### PR TITLE
Flake8 to ruff

### LINKED ISSUE #2623: Switch linting to ruff
The new kid on the block for linting is [ruff](https://github.com/charliermarsh/ruff). I tested it locally and it's much faster than flake8. Started a branch (`flake8-to-ruff`), could one of the infra whiz kids finish this? @priyanshuone6 @Saransh-cpp

### PR DESCRIPTION
**Pull request by Saransh-cpp**
# Description

Tried `ruff` locally, and it really is super fast!

I also tried migrating some of the metadata to `pyproject.toml`, but I ended up with a lot of errors (given that `pyproject.toml` support is still very new and maybe in the beta phase in `setuptools`.)

Some false alarm examples that are being ignored using `pyproject.toml` -
```
pybamm/parameters/__init__.py:5:5: F401 `.process_parameter_data.process_3D_data_csv` imported but unused

pybamm/solvers/jax_bdf_solver.py:293:9: E741 Ambiguous variable name: `I`

tests/__init__.py:15:101: E501 Line too long (106 > 100 characters)  # not a false alarm, but the import names can get lengthy sometimes in the __init__ files

pybamm/__init__.py:61:1: F403 `from .expression_tree.symbol import *` used; unable to detect undefined names

pybamm/__init__.py:235:1: E402 Module level import not at top of file
```

Fixes #2623 

## Type of change

Please add a line in the relevant section of [CHANGELOG.md](https://github.com/pybamm-team/PyBaMM/blob/develop/CHANGELOG.md) to document the change (include PR #) - note reverse order of PR #s. If necessary, also add to the list of breaking changes.

- [ ] New feature (non-breaking change which adds functionality)
- [ ] Optimization (back-end change that speeds up the code)
- [ ] Bug fix (non-breaking change which fixes an issue)

# Key checklist:

- [ ] No style issues: `$ flake8`
- [ ] All tests pass: `$ python run-tests.py --unit`
- [ ] The documentation builds: `$ cd docs` and then `$ make clean; make html`

You can run all three at once, using `$ python run-tests.py --quick`.

## Further checks:

- [ ] Code is commented, particularly in hard-to-understand areas
- [ ] Tests added that prove fix is effective or that feature works

### CODE REVIEW

**Comment by Saransh-cpp** [comment]
At the moment, `ruff` only supports `pyproject.toml` and the CLI for listing down its configuration.

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -63,17 +63,18 @@ Finally, if you really, really, _really_ love developing PyBaMM, have a look at
...
+We use [ruff](https://github.com/charliermarsh/ruff) to check our PEP8 adherence. To try this on your system, navigate to the PyBaMM directory in a console and type
 
 ```bash
-flake8
+python -m pip install pre-commit
+pre-commit run ruff
 ```
 
-Flake8 is configured inside the file `tox.ini`, under the section `[flake8]`, allowing us to ignore some errors. If you think this should be added or removed, please submit an [issue](#issues)
+ruff is configured inside the file `pre-comit.yaml`, allowing us to ignore some errors. If you think this should be added or removed, please submit an [issue](#issues)
```
Comment (Saransh-cpp) [inline CONTRIBUTING.md:75]:
```suggestion
ruff is configured inside the file `pre-commit-config.yaml`, allowing us to ignore some errors. If you think this should be added or removed, please submit an [issue](#issues)
```

**Review by tinosulzer — APPROVED** [review]
Thanks @Saransh-cpp , that was fast! I think this way with `.pre-commit-config.yaml` is a good compromise. Approval from me but will wait a bit to see if @priyanshuone6 has any suggestions

**Review by priyanshuone6 — APPROVED** [review]
Looks good! It's nice that it has an 88 character limit by default, and we don’t need to specify every time.

**Pull request merged**